compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -626,15 +626,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     }
                 };
 
-                let coerce_to = match opt_coerce_to {
-                    Some(c) => c,
-                    None => {
-                        // If the loop context is not a `loop { }`, then break with
-                        // a value is illegal, and `opt_coerce_to` will be `None`.
-                        // Return error in that case (#114529).
-                        return Ty::new_misc_error(tcx);
+                // If the loop context is not a `loop { }`, then break with
+                // a value is illegal, and `opt_coerce_to` will be `None`.
+                // Set expectation to error in that case and set tainted
+                // by error (#114529)
+                let coerce_to = opt_coerce_to.unwrap_or_else(|| {
+                    let err = Ty::new_misc_error(tcx);
+                    if let ty::Error(e) = err.kind() {
+                        self.set_tainted_by_errors(*e);
                     }
```

Comment:
Please manually delay an error instead of using an if-let.

```suggestion
                    let guar = tcx.sess.delay_span_bug(...);
                    self.set_tainted_by_errors(guar);
                    Ty::new_error(tcx, guar)
```